# Project Milestone 2
# DSC540-T303
# Ryan Krenke
# 04/21/24

## Cleaning/Formatting Flat File Source

In [3]:
import numpy as np

In [4]:
import pandas as pd

In [5]:
airdf1 = pd.read_csv("C:/Users/ryanr/dsc540/annual_conc_by_monitor_2023.csv") #reading csv file into dataframe

In [6]:
airdf1.head(50)

State Code  County Code  Site Num  Parameter Code  POC   Latitude  \
0            1            3        10           44201    1  30.497478   
1            1            3        10           44201    1  30.497478   
2            1            3        10           44201    1  30.497478   
3            1            3        10           44201    1  30.497478   
4            1            3        10           88101    3  30.497478   
5            1            3        10           88101    3  30.497478   
6            1            3        10           88101    3  30.497478   
7            1            3        10           88101    3  30.497478   
8            1            3        10           88101    3  30.497478   
9            1            3        10           88101    3  30.497478   
10           1            3        10           88101    3  30.497478   
11           1           27         1           88101    3  33.284928   
12           1           27         1           88101    3  33.284928   
13           1           27         1           88101    3  33.284928   
14           1           27         1           88101    3  33.284928   
15           1           27         1           88101    3  33.284928   
16           1           27         1           88101    3  33.284928   
17           1           27         1           88101    3  33.284928   
18           1           49      1003           88101    3  34.288567   
19           1           49      1003           88101    3  34.288567   
20           1           49      1003           88101    3  34.288567   
21           1           49      1003           88101    3  34.288567   
22           1           49      1003           88101    3  34.288567   
23           1           49      1003           88101    3  34.288567   
24           1           49      1003           88101    3  34.288567   
25           1           49      9991           44201    1  34.289001   
26           1           49      9991           44201    1  34.289001   
27           1           49      9991           44201    1  34.289001   
28           1           49      9991           44201    1  34.289001   
29           1           51         4           44201    1  32.535680   
30           1           51         4           44201    1  32.535680   
31           1           51         4           44201    1  32.535680   
32           1           51         4           44201    1  32.535680   
33           1           53      1000           61103    1  31.092100   
34           1           53      1000           61104    1  31.092100   
35           1           53      1000           62101    1  31.092100   
36           1           53      1000           62201    1  31.092100   
37           1           53      1000           63301    1  31.092100   
38           1           53      1000           64101    1  31.092100   
39           1           53      1000           65102    1  31.092100   
40           1           55        10           88101    3  33.991494   
41           1           55        10           88101    3  33.991494   
42           1           55        10           88101    3  33.991494   
43           1           55        10           88101    3  33.991494   
44           1           55        10           88101    3  33.991494   
45           1           55        10           88101    3  33.991494   
46           1           55        10           88101    3  33.991494   
47           1           55        11           44201    1  33.904039   
48           1           55        11           44201    1  33.904039   
49           1           55        11           44201    1  33.904039   

    Longitude  Datum              Parameter Name          Sample Duration  \
0  -87.880258  NAD83                       Ozone                   1 HOUR   
1  -87.880258  NAD83                       Ozone  8-HR RUN AVG BEGIN HOUR   
2  -87.880258  NAD83                       Ozone 

In [7]:
len(airdf1) 

51258

My flat file has 51,258 rows. The data is comprised of data from many air quality sampling sites all around the United States and Mexico. 

### Step 1 

My first action is to understand if my dataset has any blank values. If there are, I will need to determine how to handle them. 

In [8]:
airdf1.isnull().sum() #using isnull function to identify blank cells and using sum function to get total blanks for each column. 

State Code                           0
County Code                          0
Site Num                             0
Parameter Code                       0
POC                                  0
Latitude                             0
Longitude                            0
Datum                                0
Parameter Name                       0
Sample Duration                      0
Pollutant Standard               32828
Metric Used                          0
Method Name                      10478
Year                                 0
Units of Measure                     0
Event Type                           0
Observation Count                    0
Observation Percent                  0
Completeness Indicator               0
Valid Day Count                      0
Required Day Count                   0
Exceptional Data Count               0
Null Data Count                      0
Primary Exceedance Count         38772
Secondary Exceedance Count       39602
Certification Indicator  

### Step 2

After analyzing which columns contain blank values, I think I am fortunate that I will not need to use those columns. I will make a new dataframe with just the columns I am interested in. 

In [9]:
airdf2 = airdf1[['State Code', 'Site Num', 'Parameter Code', 'Latitude', 'Longitude', 'Parameter Name', 'Units of Measure', 'Observation Count', 'Observation Percent', 'State Name']]

In [10]:
airdf2.head(20)

State Code  Site Num  Parameter Code   Latitude  Longitude  \
0            1        10           44201  30.497478 -87.880258   
1            1        10           44201  30.497478 -87.880258   
2            1        10           44201  30.497478 -87.880258   
3            1        10           44201  30.497478 -87.880258   
4            1        10           88101  30.497478 -87.880258   
5            1        10           88101  30.497478 -87.880258   
6            1        10           88101  30.497478 -87.880258   
7            1        10           88101  30.497478 -87.880258   
8            1        10           88101  30.497478 -87.880258   
9            1        10           88101  30.497478 -87.880258   
10           1        10           88101  30.497478 -87.880258   
11           1         1           88101  33.284928 -85.803608   
12           1         1           88101  33.284928 -85.803608   
13           1         1           88101  33.284928 -85.803608   
14           1         1           88101  33.284928 -85.803608   
15           1         1           88101  33.284928 -85.803608   
16           1         1           88101  33.284928 -85.803608   
17           1         1           88101  33.284928 -85.803608   
18           1      1003           88101  34.288567 -85.969858   
19           1      1003           88101  34.288567 -85.969858   

              Parameter Name             Units of Measure  Observation Count  \
0                      Ozone            Parts per million               2660   
1                      Ozone            Parts per million               2769   
2                      Ozone            Parts per million               2769   
3                      Ozone            Parts per million               1963   
4   PM2.5 - Local Conditions  Micrograms/cubic meter (LC)               4005   
5   PM2.5 - Local Conditions  Micrograms/cubic meter (LC)                166   
6   PM2.5 - Local Conditions  Micrograms/cubic meter (LC)                166   
7   PM2.5 - Local Conditions  Micrograms/cubic meter (LC)                166   
8   PM2.5 - Local Conditions  Micrograms/cubic meter (LC)                166   
9   PM2.5 - Local Conditions  Micrograms/cubic meter (LC)                166   
10  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)                166   
11  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)               3750   
12  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)                155   
13  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)                155   
14  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)                155   
15  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)                155   
16  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)                155   
17  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)                155   
18  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)               4310   
19  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)                181   

    Observation Percent State Name  
0                    48    Alabama  
1                    47    Alabama  
2                    47    Alabama  
3                    46    Alabama  
4                    46    Alabama  
5                    45    Alabama  
6                    45    Alabama  
7                    45    Alabama  
8                    45    Alabama  
9                    45    Alabama  
10                   45    Alabama  
11                   43    Alabama  
12                   42    Alabama  
13                   42    Alabama  
14                   42    Alabama  
15                   42    Alabama  
16                   42    Alabama  
17                   42    Alabama  
18                   49    Alabama  
19                   50    Alabama

### Step 3

I am interested to learn more about the air testing methods, so I will use the groupby function to group the dataframe by Parameter Name and then look at the mean results and then some details of Observation Count (results) with describe. 

In [11]:
airbyparameter = airdf2.groupby('Parameter Name') #grouping dataframe by Parameter Name column
airbyparameter.mean() #calling for the mean of each numerical field

C:\Users\ryanr\AppData\Local\Temp\ipykernel_22448\923359952.py:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  airbyparameter.mean() #calling for the mean of each numerical field


State Code     Site Num  \
Parameter Name                                                   
1,1,1,2,2-Pentafluoroethane              6.000000   916.454545   
1,1,1,2-Tetrachloroethane               31.250000   872.625000   
1,1,2,2-Tetrachloroethane               32.141593   797.044248   
1,1,2-Trichloro-1,2,2-trifluoroethane   27.056604  1012.358491   
1,1,2-Trichloroethane                   32.123894   766.017699   
...                                           ...          ...   
trans-1,2-Dichloroethylene              32.246914   798.851852   
trans-1,3-Dichloropropene               29.327586   861.689655   
trans-2-Butene                          30.312500   947.593750   
trans-2-Hexene                          26.307692  1452.307692   
trans-2-Pentene                         32.518519  1084.666667   

                                       Parameter Code   Latitude   Longitude  \
Parameter Name                                                                 
1,1,1,2,2-Pentafluoroethane                   43834.0  35.983881 -119.492646   
1,1,1,2-Tetrachloroethane                     43837.0  42.573333  -75.280225   
1,1,2,2-Tetrachloroethane                     43818.0  39.440407  -84.555559   
1,1,2-Trichloro-1,2,2-trifluoroethane         43821.0  38.092999  -94.436510   
1,1,2-Trichloroethane                         43820.0  39.588348  -84.914016   
...                                               ...        ...         ...   
trans-1,2-Dichloroethylene                    43838.0  40.369772  -86.199244   
trans-1,3-Dichloropropene                     43830.0  39.385026  -88.209984   
trans-2-Butene                                43216.0  34.719534  -96.462104   
trans-2-Hexene                                43289.0  36.858971 -101.911692   
trans-2-Pentene                               43226.0  35.460645  -96.726650   

                                       Observation Count  Observation Percent  
Parameter Name                                                                 
1,1,1,2,2-Pentafluoroethane                     9.454545            17.090909  
1,1,1,2-Tetrachloroethane                      17.000000            28.250000  
1,1,2,2-Tetrachloroethane                      18.823009            33.955752  
1,1,2-Trichloro-1,2,2-trifluoroethane          16.622642            31.396226  
1,1,2-Trichloroethane                          18.504425            33.415929  
...                                                  ...                  ...  
trans-1,2-Dichloroethylene                     19.197531            35.209877  
trans-1,3-Dichloropropene                      20.077586            35.991379  
trans-2-Butene                               1417.812500            31.593750  
trans-2-Hexene                                 20.153846            20.769231  
trans-2-Pentene                              1617.555556            29.407407  

[436 rows x 7 columns]

In [12]:
airbyparameter.describe()['Observation Count'] #using describe function for statistical details of the Observation Count column

count         mean          std  min  \
Parameter Name                                                                
1,1,1,2,2-Pentafluoroethane             11.0     9.454545     1.439697  6.0   
1,1,1,2-Tetrachloroethane               16.0    17.000000     6.033241  7.0   
1,1,2,2-Tetrachloroethane              113.0    18.823009     9.242014  1.0   
1,1,2-Trichloro-1,2,2-trifluoroethane   53.0    16.622642    11.042273  1.0   
1,1,2-Trichloroethane                  113.0    18.504425     9.157475  1.0   
...                                      ...          ...          ...  ...   
trans-1,2-Dichloroethylene              81.0    19.197531     9.996774  1.0   
trans-1,3-Dichloropropene              116.0    20.077586     7.809861  3.0   
trans-2-Butene                          32.0  1417.812500  2061.026106  5.0   
trans-2-Hexene                          13.0    20.153846     9.467894  5.0   
trans-2-Pentene                         27.0  1617.555556  2179.986544  3.0   

                                         25%   50%     75%     max  
Parameter Name                                                      
1,1,1,2,2-Pentafluoroethane             9.00  10.0    10.0    11.0  
1,1,1,2-Tetrachloroethane              14.75  15.0    17.5    27.0  
1,1,2,2-Tetrachloroethane              14.00  16.0    28.0    35.0  
1,1,2-Trichloro-1,2,2-trifluoroethane   7.00  16.0    28.0    35.0  
1,1,2-Trichloroethane                  14.00  16.0    28.0    35.0  
...                                      ...   ...     ...     ...  
trans-1,2-Dichloroethylene             13.00  20.0    28.0    35.0  
trans-1,3-Dichloropropene              15.00  20.0    28.0    35.0  
trans-2-Butene                         23.75  30.0  3799.5  5511.0  
trans-2-Hexene                         15.00  15.0    29.0    30.0  
trans-2-Pentene                        15.00  26.0  4162.0  5511.0  

[436 rows x 8 columns]

In [13]:
airdf2['Parameter Name'].nunique() #using nunique function to confirm the number of different methods

436

I was guessing while glancing at the file over the past weeks that there were only a couple methods (Parameter Name), but I have quickly realized there are more than 400. I don't know a lot about air quality testing, but quick research tells me that I can't compare the results (Observation Count) for all of these different methods as being equal. I will need to limit them somehow. 

### Step 4

I am investigating ways to limit the number of methods in my dataset. 

In [14]:
airbyparameter2 = airdf2.sort_values('Parameter Name') #sorting dataframe by Parameter Name column

In [15]:
airbyparameter2

State Code  Site Num  Parameter Code   Latitude   Longitude  \
5536            6      8001           43834  33.999580 -117.416010   
7295            6         5           43834  37.348497 -121.894898   
5196            6         6           43834  38.746430 -121.264980   
6873            6      1003           43834  37.961578 -121.281414   
3923            6        14           43834  35.356615 -119.062613   
...           ...       ...             ...        ...         ...   
8062            8         9           43226  39.985105 -104.866658   
9303            8         8           43226  40.209387 -104.824050   
42911          48      1044           43226  32.175417  -96.870189   
36396          40       522           43226  36.926603  -94.873439   
45230          48        17           43226  27.501826  -99.502984   

                    Parameter Name          Units of Measure  \
5536   1,1,1,2,2-Pentafluoroethane  Parts per billion Carbon   
7295   1,1,1,2,2-Pentafluoroethane  Parts per billion Carbon   
5196   1,1,1,2,2-Pentafluoroethane  Parts per billion Carbon   
6873   1,1,1,2,2-Pentafluoroethane  Parts per billion Carbon   
3923   1,1,1,2,2-Pentafluoroethane  Parts per billion Carbon   
...                            ...                       ...   
8062               trans-2-Pentene  Parts per billion Carbon   
9303               trans-2-Pentene  Parts per billion Carbon   
42911              trans-2-Pentene  Parts per billion Carbon   
36396              trans-2-Pentene  Parts per billion Carbon   
45230              trans-2-Pentene  Parts per billion Carbon   

       Observation Count  Observation Percent  State Name  
5536                  10                   16  California  
7295                  10                   16  California  
5196                   9                   15  California  
6873                  10                   16  California  
3923                   6                   10  California  
...                  ...                  ...         ...  
8062                  14                    3    Colorado  
9303                  21                    4    Colorado  
42911                 15                   25       Texas  
36396                  9                   23    Oklahoma  
45230                 15                   25       Texas  

[51258 rows x 10 columns]

I can take a random sample of my data to reduce the number of methods. I will sample 1000 random rows from my dataset. 

In [16]:
air2subset = airdf2.sample(1000, random_state=14) # sample function to sample 1000 random rows of data

In [17]:
air2subset

State Code  Site Num  Parameter Code   Latitude   Longitude  \
24577          28        20           88101  32.329111  -90.182722   
33056          38         3           88101  46.825425 -100.768210   
35002          39        17           43231  40.366300  -80.615800   
20992          24        30           68114  39.055277  -76.878333   
40088          42      5200           43812  40.268963  -80.243995   
...           ...       ...             ...        ...         ...   
22966          26         1           88115  42.228620  -83.208200   
24782          29         5           44201  39.303174  -94.377014   
29158          35        23           68116  35.134300 -106.585200   
10354          12        34           43207  26.053889  -80.256944   
38552          42         7           14129  40.673656  -80.317731   

                      Parameter Name             Units of Measure  \
24577       PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
33056       PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
35002                       n-Hexane     Parts per billion Carbon   
20992  Sample Volume - Teflon Filter                  Cubic meter   
40088                   Chloroethane     Parts per billion Carbon   
...                              ...                          ...   
22966              Chlorine PM2.5 LC  Micrograms/cubic meter (LC)   
24782                          Ozone            Parts per million   
29158  Sample Volume - Quartz Filter                  Cubic meter   
10354                      Freon 113     Parts per billion Carbon   
38552                  Lead (TSP) LC  Micrograms/cubic meter (LC)   

       Observation Count  Observation Percent    State Name  
24577                171                   47   Mississippi  
33056                125                   34  North Dakota  
35002                 35                  100          Ohio  
20992                 40                   33      Maryland  
40088                 16                   27  Pennsylvania  
...                  ...                  ...           ...  
22966                 40                   33      Michigan  
24782               2127                   36      Missouri  
29158                 40                   33    New Mexico  
10354                 30                   50       Florida  
38552                 27                   45  Pennsylvania  

[1000 rows x 10 columns]

In [18]:
air2subset['Parameter Name'].nunique() # Number of unique methods in subset

238

In [19]:
air2subset['State Name'].nunique() # Number of unique states in subset

50

With still 238 unique methods, I'd like to understand how often each are used in the data. I will use value_counts to count how many times each method is used. 

In [20]:
airdf2.value_counts('Parameter Name') #value_counts function to count how many times each Parameter Name is used

Parameter Name
PM2.5 - Local Conditions            10386
Ozone                                5176
Sulfur dioxide                       1808
PM10 Total 0-10um STP                1237
Nitrogen dioxide (NO2)                862
                                    ...  
Light scatter (miles visibility)        1
Light scatter (ug/m3)                   1
Black Carbon  PM2.5 Corrected           1
Total Carbon PM2.5 LC TOT               1
Soil index (COH)                        1
Length: 436, dtype: int64

### Step 5

This is still several hundred methods to deal with. After additional research, it appears like the methods referring to "PM" for "Particulate Matter" are common measures of air quality. I recognize two of these methods in the top 5 most common methods used in my dataset, so I will create a new dataframe with just these two methods. 

In [21]:
# new dataframe using two Parameter Name values from original dataframe
airdf3 = airdf2[(airdf2['Parameter Name'] == 'PM2.5 - Local Conditions') | (airdf2['Parameter Name'] == 'PM10 Total 0-10um STP')]

In [22]:
airdf3

State Code  Site Num  Parameter Code   Latitude   Longitude  \
4               1        10           88101  30.497478  -87.880258   
5               1        10           88101  30.497478  -87.880258   
6               1        10           88101  30.497478  -87.880258   
7               1        10           88101  30.497478  -87.880258   
8               1        10           88101  30.497478  -87.880258   
...           ...       ...             ...        ...         ...   
51249          80         6           88101  31.291293 -110.951513   
51250          80         6           88101  31.291293 -110.951513   
51251          80         6           88101  31.291293 -110.951513   
51252          80         6           88101  31.291293 -110.951513   
51253          80         6           88101  31.291293 -110.951513   

                 Parameter Name             Units of Measure  \
4      PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
5      PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
6      PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
7      PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
8      PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
...                         ...                          ...   
51249  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
51250  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
51251  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
51252  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
51253  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   

       Observation Count  Observation Percent         State Name  
4                   4005                   46            Alabama  
5                    166                   45            Alabama  
6                    166                   45            Alabama  
7                    166                   45            Alabama  
8                    166                   45            Alabama  
...                  ...                  ...                ...  
51249                268                   73  Country Of Mexico  
51250                268                   73  Country Of Mexico  
51251                268                   73  Country Of Mexico  
51252                268                   73  Country Of Mexico  
51253                268                   73  Country Of Mexico  

[11623 rows x 10 columns]

In [23]:
airdf3['State Name'].nunique() # number of unique states in new dataframe

52

In [24]:
uniquestates = set(airdf3['State Name']) #creating a set with all the unique states in the dataframe

In [25]:
uniquestates

{'Alabama',
 'Alaska',
 'Arizona',
 'Arkansas',
 'California',
 'Colorado',
 'Connecticut',
 'Country Of Mexico',
 'District Of Columbia',
 'Florida',
 'Georgia',
 'Hawaii',
 'Idaho',
 'Illinois',
 'Indiana',
 'Iowa',
 'Kansas',
 'Kentucky',
 'Louisiana',
 'Maine',
 'Maryland',
 'Massachusetts',
 'Michigan',
 'Minnesota',
 'Mississippi',
 'Missouri',
 'Montana',
 'Nebraska',
 'Nevada',
 'New Hampshire',
 'New Jersey',
 'New Mexico',
 'New York',
 'North Carolina',
 'North Dakota',
 'Ohio',
 'Oklahoma',
 'Oregon',
 'Pennsylvania',
 'Puerto Rico',
 'Rhode Island',
 'South Carolina',
 'South Dakota',
 'Tennessee',
 'Texas',
 'Utah',
 'Vermont',
 'Virginia',
 'Washington',
 'West Virginia',
 'Wisconsin',
 'Wyoming'}

### Step 6

I am only interested in locations within the United States, so looking over the list of unique states, I will need to remove the Country of Mexico. 

In [26]:
airdf4 = airdf3[(airdf3['State Name'] != 'Country Of Mexico')] # creating new dataframe that does not include Country of Mexico

In [27]:
airdf4 

State Code  Site Num  Parameter Code   Latitude  Longitude  \
4               1        10           88101  30.497478 -87.880258   
5               1        10           88101  30.497478 -87.880258   
6               1        10           88101  30.497478 -87.880258   
7               1        10           88101  30.497478 -87.880258   
8               1        10           88101  30.497478 -87.880258   
...           ...       ...             ...        ...        ...   
51231          72         4           88101  18.009558 -66.627249   
51232          72         4           88101  18.009558 -66.627249   
51233          72         4           88101  18.009558 -66.627249   
51234          72         4           88101  18.009558 -66.627249   
51235          72         4           88101  18.009558 -66.627249   

                 Parameter Name             Units of Measure  \
4      PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
5      PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
6      PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
7      PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
8      PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
...                         ...                          ...   
51231  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
51232  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
51233  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
51234  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
51235  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   

       Observation Count  Observation Percent   State Name  
4                   4005                   46      Alabama  
5                    166                   45      Alabama  
6                    166                   45      Alabama  
7                    166                   45      Alabama  
8                    166                   45      Alabama  
...                  ...                  ...          ...  
51231                 50                   41  Puerto Rico  
51232                 50                   41  Puerto Rico  
51233                 50                   41  Puerto Rico  
51234                 50                   41  Puerto Rico  
51235                 50                   41  Puerto Rico  

[11614 rows x 10 columns]

### Step 7

The dataframe index is still showing the random values. I will need to reset the index to make it cleaner to read and use further. 

In [28]:
airdf4.reset_index(drop=True) #using reset_index function to reset the index of the dataframe in order starting at 0

State Code  Site Num  Parameter Code   Latitude  Longitude  \
0               1        10           88101  30.497478 -87.880258   
1               1        10           88101  30.497478 -87.880258   
2               1        10           88101  30.497478 -87.880258   
3               1        10           88101  30.497478 -87.880258   
4               1        10           88101  30.497478 -87.880258   
...           ...       ...             ...        ...        ...   
11609          72         4           88101  18.009558 -66.627249   
11610          72         4           88101  18.009558 -66.627249   
11611          72         4           88101  18.009558 -66.627249   
11612          72         4           88101  18.009558 -66.627249   
11613          72         4           88101  18.009558 -66.627249   

                 Parameter Name             Units of Measure  \
0      PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
1      PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
2      PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
3      PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
4      PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
...                         ...                          ...   
11609  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
11610  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
11611  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
11612  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
11613  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   

       Observation Count  Observation Percent   State Name  
0                   4005                   46      Alabama  
1                    166                   45      Alabama  
2                    166                   45      Alabama  
3                    166                   45      Alabama  
4                    166                   45      Alabama  
...                  ...                  ...          ...  
11609                 50                   41  Puerto Rico  
11610                 50                   41  Puerto Rico  
11611                 50                   41  Puerto Rico  
11612                 50                   41  Puerto Rico  
11613                 50                   41  Puerto Rico  

[11614 rows x 10 columns]

### Fully cleaned and transformed flat file dataset

In [29]:
airdf4 

State Code  Site Num  Parameter Code   Latitude  Longitude  \
4               1        10           88101  30.497478 -87.880258   
5               1        10           88101  30.497478 -87.880258   
6               1        10           88101  30.497478 -87.880258   
7               1        10           88101  30.497478 -87.880258   
8               1        10           88101  30.497478 -87.880258   
...           ...       ...             ...        ...        ...   
51231          72         4           88101  18.009558 -66.627249   
51232          72         4           88101  18.009558 -66.627249   
51233          72         4           88101  18.009558 -66.627249   
51234          72         4           88101  18.009558 -66.627249   
51235          72         4           88101  18.009558 -66.627249   

                 Parameter Name             Units of Measure  \
4      PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
5      PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
6      PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
7      PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
8      PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
...                         ...                          ...   
51231  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
51232  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
51233  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
51234  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   
51235  PM2.5 - Local Conditions  Micrograms/cubic meter (LC)   

       Observation Count  Observation Percent   State Name  
4                   4005                   46      Alabama  
5                    166                   45      Alabama  
6                    166                   45      Alabama  
7                    166                   45      Alabama  
8                    166                   45      Alabama  
...                  ...                  ...          ...  
51231                 50                   41  Puerto Rico  
51232                 50                   41  Puerto Rico  
51233                 50                   41  Puerto Rico  
51234                 50                   41  Puerto Rico  
51235                 50                   41  Puerto Rico  

[11614 rows x 10 columns]

The dataset that I chose to use was already a fairly clean dataset. The only thing that really needed to be cleaned up from the start was the blank values. Fortunately, the data columns that I was interested in did not contain any blank values, so I could just include those columns. 

The other main issue to address was that my dataset was very large and, because it has hundreds of different testing methods that couldn't be compared equally, I needed to reduce it in an appropriate way. Therefore, I had to spend some time understanding the different characteristics and trialing ways to reduce. One ethical implication I came across was when I chose to pick a random sample from the data. I felt a little uncomfortable about doing this because I couldn't be sure about what data I was or was not including. However, I believe I chose a large enough sample size to still make it representative of the whole dataset, and in the end, I did not end up using this method in my final dataset anyway.

After further research, I was able to reduce the data to just two methods that could be compared better. Then I just needed to make sure the data only contained locations from within the United States. There is a slight risk from this choice as my data does not contain data from one state: Delaware. Honestly, I did not even check to see if the original data included Delaware, but I feel this is a pretty small state and there should be plenty of data from nearby states to offset this risk. 

As far as data credibility and sourcing, I have no concerns with these. The data is openly shared by a credible resource - the United States EPA (Environmental Protection Agency). The data collection process was performed using standard testing methods from known, identified sources. I do recognize that the EPA is a government agency, therefore this data could be considered regulatory data, so I will have to make sure I am reporting accurate results. 


In [34]:
import csv

In [37]:
airdf4.to_csv(r'C:\Users\ryanr\dsc540\air4csv', index = False)